numerics/root_finders_test.cpp

```diff
@@ -54,7 +57,134 @@ TEST_F(RootFindersTest, SquareRoots) {
+  // this is because that one is overly close to the machine precision, so that
+  // attempting to perform as many interpolation steps as theoretically possible
+  // before the bisection leads to early bisection in practice, because the
+  // interval shrinkage becomes too great as the convergent values get
+  // discretized.
+  double const δ = 0x1p-50;
+  // The factor p theoretically only needs to be larger than √2, but, for
+  // similar reasons, bringing it closer (e.g., 17/12) requires increasing δ,
+  // which overall reduces the number of evaluations required.
+  double const p = 1.5;
+
+  X.push_back(b);
+  int bisections = 0;
+  for (; a < X.back(); ++bisections) {
+    // k is the number of interpolation steps on the interval [a, X.back()]
+    // before a bisection happens again.
+    int const k = std::round(std::log2((X.back() - a) / δ));
+    for (int j = 1; j <= k; ++j) {
+      X.push_back(X.back() - std::pow(p, k - j) * δ);
+    }
+    expected_bisections.insert(X.size() - 1);
+    X.push_back(a + (X.back() - a) / 2);
   }
+
+  // [WG13], equation (5).
+  auto const s = [](double xa, double xc, double xb, double fa) {
+    return fa * (xb - xc) / (xa - xc);
+  };
+
+  // [WG13], equations (7) and (9).
+  auto const q =
+      [](double xa, double xd, double xc, double xb, double fa, double fb) {
+        double const α = (xa - xd) * fb + (xd - xb) * fa;
+        double const β = (xb - xd) * Pow<2>(fa) + (xd - xa) * Pow<2>(fb);
+        double const γ = fa * fb * (fb - fa) * (xc - xd);
+        return -2 * γ / (β + Sqrt(Pow<2>(β) - 4 * α * γ));
+      };
+
+  // Capture everything but the function itself by copy so we have some
+  // confidence that the function is pure.
+  // We capture f because it is defined recursively.
+  // If non-null, |*evaluations| is incremented when the function is called.
+  // If |expect_brent_calls| is true, this function checks that x is an element
+  // of the expected sequence X.
+  std::function<double(double, int*, bool)> f =
+      [&f, s, q, a, b, X, expected_bisections](
+          double const x,
+          int* const evaluations,
+          bool const expect_brent_calls) -> double {
+    double const f_a = -100;
+    if (evaluations != nullptr) {
+      ++*evaluations;
+    }
+    if (x == a) {
+      return f_a;
+    }
+    if (x == b) {
+      return s(a, X[1], b, f_a);
+    }
+    // [WG13] define f only on X. We extend it as a step function, returning the
+    // value for the element of X nearest to the given x.
+    int j;
+    double min_Δx = std::numeric_limits<double>::infinity();
+    for (int i = 1; i < X.size() - 1; ++i) {
+      double const Δx = std::abs(x - X[i]);
+      if (Δx < min_Δx) {
+        min_Δx = Δx;
+        j = i;
+      }
+    }
+    if (expect_brent_calls) {
+      EXPECT_THAT(x, AlmostEquals(X[j], 0)) << j;
+    }
+
+    if (expected_bisections.count(j) != 0) {
+      return 100;
```

Comment:
I have no idea what this is doing.  A comment maybe?  If this is "any random value", consider using `NaN`.

Reply:
From the paper:
> Since we are bisecting the function regardless of its value at Xk , we may choose an arbitrary positive value for f (Xk ), such as 100.